File: 844_BackSpace_String_Compare.java
/*
Runtime: 9 ms, faster than 5.39% of Java online submissions for Backspace String Compare.
Memory Usage: 40.2 MB, less than 5.02% of Java online submissions for Backspace String Compare.
*/ 

class Solution {
    
    public String backspaceCompareHelper(String S)
    {
        Stack stackk = new Stack<Character>(); 
        for(int i = 0; i < S.length(); i++)
        {
            char character = S.charAt(i); 
            if(character == '#')
                if(stackk.size() > 0) stackk.pop(); 
                else
                {
                    
                }
            else
                {
                    stackk.add(character); 
                }
        }
        System.out.println(stackk.toString()); 
        return stackk.toString(); 
    }
    
    public boolean backspaceCompare(String S, String T) {
        return backspaceCompareHelper(S).equals( backspaceCompareHelper(T)); 
    }
}

File: 844_BackSpace_String_Compare.py
'''
Runtime: 36 ms, faster than 46.88% of Python3 online submissions for Backspace String Compare.
Memory Usage: 13.8 MB, less than 74.66% of Python3 online submissions for Backspace 

Time: O(M+N) where M, N are the lengths of S and T 
Space: O(M + N)
'''
class Solution:
    def createStack(self, S): 
        stack = [] 
        for i in S:
            if i == "#" and len(stack) > 0:
                stack.pop()
            elif i != "#":
                stack.append(i) 
            else:
                pass 
        return stack 
    
    def backspaceCompare(self, S: str, T: str) -> bool:
        return self.createStack(S) == self.createStack(T)
        